File: restricted_boltzmann_machine/cpp_codes_new/rbm.cpp
#include "rbm.hpp"


RBM::RBM(int n_visible, int n_hidden, double eta, double mom){
    n_hidden_ = n_hidden;
    n_visible_ = n_visible;
    eta_ = eta;
    mom_ = mom;

    weights_ = arma::randn<arma::mat>(n_visible_, n_hidden_)*0.01;
    visible_bias_ = arma::randn<arma::vec>(n_visible_)*0.01;
    hidden_bias_ = arma::randn<arma::vec>(n_hidden_)*0.01;


    visible_act_ = arma::vec(n_visible_).fill(0.);
    hidden_act_ = arma::vec(n_hidden_).fill(0.);
    dW_ = arma::mat(n_visible_, n_hidden_).fill(0.);
    dvb_ = arma::vec(n_visible_).fill(0.);
    dhb_ = arma::vec(n_hidden_).fill(0.);

    std::cout << "dW rows = " << dW_.n_rows << std::endl;
    std::cout << "dW cols = " << dW_.n_cols << std::endl;
}

arma::vec RBM::sigmoid(arma::vec x){
    return 1./(1 + arma::exp(-x));
}

void RBM::visible_act(){
    arma::vec u = arma::randu(n_visible_);
    visible_prob_ = sigmoid(visible_bias_ + weights_*hidden_act_);
    visible_act_ = visible_prob_ - u;
    visible_act_.transform( [](double val){return (val >= 0);} );
}

void RBM::hidden_act(){
    arma::vec u = arma::randu(n_hidden_);
    hidden_prob_ = sigmoid(hidden_bias_ + weights_.t()*visible_act_);
    hidden_act_ = hidden_prob_ - u;
    hidden_act_.transform( [](double val){return (val >= 0);} );
}

void RBM::fit(arma::mat data, int epochs, int batch_sz, int nCDsteps){
    int n_rows = data.n_rows;
    int n_cols = data.n_cols;
    nCDsteps_ = nCDsteps;

    std::default_random_engine gen;
    std::uniform_int_distribution<int> dist(0,n_cols-1);
    int batch_sz_inv = 1./batch_sz;

    for (int epoch = 0; epoch < epochs; epoch++){
        std::cout << "epoch: " << epoch << " of " << epochs << std::endl;
        for (int b = 0; b < batch_sz; b++){
            int idx = dist(gen);
            arma::vec x = data.col(idx);

            visible_act_.swap(x);
            hidden_act();

            //Positive phase
            arma::mat CDpos = visible_act_*hidden_prob_.t(); //Outer product
            arma::vec CDpos_vb = visible_act_;
            arma::vec CDpos_hb = hidden_prob_;


            for (int j = 0; j < nCDsteps; j++){
                visible_act();
                hidden_act();
            }

            arma::mat CDneg = visible_act_*hidden_prob_.t(); //outer product u*v^T
            arma::vec CDneg_vb = visible_act_;
            arma::vec CDneg_hb = hidden_prob_;

            dW_ += eta_*(CDpos - CDneg);
            dvb_ += eta_*(CDpos_vb - CDneg_vb);
            dhb_ += eta_*(CDpos_hb - CDneg_hb);

        }
        dW_ *= batch_sz_inv;
        dvb_ *= batch_sz_inv;
        dhb_ *= batch_sz_inv;

        weights_ += dW_;
        visible_bias_ += dvb_;
        hidden_bias_ += dhb_;

        dW_.fill(0.);
        dvb_.fill(0.);
        dhb_.fill(0.);
    }
}

arma::vec RBM::predict(arma::vec x){
    visible_act_ = x;
    hidden_act();
    visible_act();
    for (int k = 0; k < nCDsteps_; k++){
        hidden_act();
        visible_act();
    }

    return visible_act_;
}

File: restricted_boltzmann_machine/python/rbm.py
#Necessary imports.
import numpy as np
import matplotlib.pyplot as plt
from progress.bar import Bar, IncrementalBar

class rbm:
    """
    This class implement a simple restricted boltzmann machine (RBM) using both learning rate (eta) and momentum to update its parameters;
    weights, visiblebias and hiddenbias, to learn the RBM features contained in the sample distribution it's supposed to learn.
    """
    def __init__(self, nvisible, nhidden, eta, momentum, nCDsteps, nepochs, batch_size, size_of_dataset):
        """
        Implements a simple Restricted Boltzmann Machine.


        nvisible: number of visible nodes
        nhidden: number of hidden nodes.
        eta: Learning rate; set 0.1 <= eta <= 0.5
        momentum: set 0.1 <= momentum <= 0.9.
        nCDsteps: number of iterations in the contrastive divergence algorithm.
        nepochs: number of epochs
        """
        self.nvisible = nvisible
        self.nhidden = nhidden
        self.eta = eta
        self.momentum = momentum                                        #Parameter in the learning rule.
        self.nCDsteps = nCDsteps                                        #Number of CD-steps.
        self.nepochs = nepochs                                          #Number of epochs per training examples.
        self.batch_size = batch_size
        self.size_of_dataset = size_of_dataset
        self.visibleprob = np.zeros(nvisible)                           #Stores an empty array for the visible probabilities.
        self.visibleact = np.zeros(nvisible)                            #Stores an empty array for the activations of the visible nodes.
        self.hiddenprob = np.zeros(nhidden)                             #Stores an empty array for the hidden probabilities.
        self.hiddenact = np.zeros(nhidden)                              #Stores an empty array for the hidden activations.

        #Defines the parameters in the model.
        self.weights = np.random.randn(nvisible, nhidden)*0.01          #Initial weights
        self.visiblebias = np.random.randn(nvisible)*0.01               #Initial visible bias
        self.hiddenbias = np.random.randn(nhidden)*0.01                 #Initial hidden bias
        self.loss = np.zeros(self.nepochs)
        self.data_matrix = np.zeros((self.size_of_dataset, self.nvisible))

    def sigmoid(self, x):
        """Activation function"""
        return 1./(1. + np.exp(-x))


    def compute_visible(self,hidden):
        """Computes P = p(v[i]|h) = 1 for i = 1,2,...,nvisible, and activates node i given that P > u,
        where u is a random number between 0 and 1 sampled from a uniform distribution
        -----------------------------------------Variables-------------------------------------------
        hidden: an array containing the activations of the hidden nodes.
        """
        u = np.random.uniform(0,1.0,self.nvisible)
        self.visibleprob = self.sigmoid(self.visiblebias + np.dot(self.weights, hidden))
        self.visibleact = self.visibleprob > u
        return [self.visibleprob, self.visibleact]

    def compute_hidden(self, visible):
        """Computes P = p(h[j]|v) = 1 for j = 1,2,...,nhidden, and activates node j given that p > u,

        -----------------------------------------Variables-------------------------------------------
        visible: array containing the activations of the visible nodes.
        """
        u = np.random.uniform(0,1,self.nhidden)
        self.hiddenprob = self.sigmoid(self.hiddenbias + np.dot(self.weights.T, visible))
        self.hiddenact = self.hiddenprob > u
        return [self.hiddenprob, self.hiddenact]


    def train_model(self, input_data):
        """
        An implementation of the CD-n algorithm.


        inputs: training data should be a vector V of the same shape as v = np.zeros(nvisible)
        """

        self.data_matrix = input_data
        N = self.nvisible                                    #Scaling factor used in the learning process.
        #Creates empty arrays to store "differentials".
        dW = np.zeros((self.nvisible,self.nhidden))
        dvb = np.zeros(self.nvisible)
        dhb = np.zeros(self.nhidden)

        bar = IncrementalBar("Progress", max = self.nepochs)  #Sets up the progressbar.
        #Trains the RBM using the CD-n algorithm on a single datapoint at the time.
        for epoch in range(self.nepochs):
            bar.next()
            shuffled_indices = np.random.permutation(self.batch_size)
            training_data = self.data_matrix[shuffled_indices]
            error = 0
            for k in range(self.batch_size):
                visible = training_data[k]
                #sample hidden variables
                self.compute_hidden(visible)

                #compute <vh>_0
                CDpos = np.tensordot(visible, self.hiddenprob, axes = 0)     #Tensor product computes a matrix of shape (nvisible x nhidden)
                CDpos_vb = visible                                          #Simply the initial state of the visible nodes.
                CDpos_hb = self.hiddenprob                                  #The first computed state of the hidden nodes.

                #CD-n, if nCDsteps = 1, this is essentially just reconstrunction of the input.
                #Choosing nCDsteps = 1 works alright and is computationally effective.
                for j in range(self.nCDsteps):
                    self.compute_visible(self.hiddenact)
                    self.compute_hidden(self.visibleact)
                #self.compute_visible(self.hiddenprob)
                #self.compute_hidden(self.visibleact)

                #Computes <vh>_n
                CDneg = np.tensordot(self.visibleact, self.hiddenprob, axes = 0)
                CDneg_vb = self.visibleact
                CDneg_hb = self.hiddenprob

                #This is where the learning happens, you can skip the momentum if you want but it speeds up initial learning
                #You can modifiy the class to add decay, that is add -self.decay*dW to the learning rule, or reduce the momentum towards the end of learning.

                #Reconstruction error. It measures how well the RBM reconstructs the data it's shown.
                visible = training_data[k]
                error += np.sum((self.data_matrix[k]-self.visibleact)**2)
                dW = self.eta*(CDpos - CDneg)/N + self.momentum*dW
                self.weights += dW
                dvb = self.eta*(CDpos_vb - CDneg_vb)/N + self.momentum*dvb
                self.visiblebias += dvb
                dhb = self.eta*(CDpos_hb - CDneg_hb)/N + self.momentum*dhb
                self.hiddenbias += dhb
            error /= self.batch_size
            self.loss[epoch] = error
        bar.finish()


    def predict(self, input):
        self.compute_hidden(input)
        self.compute_visible(self.hiddenprob)
        for k in range(self.nCDsteps):
            self.compute_hidden(self.visibleact)
            self.compute_visible(self.hiddenprob)

    def plot_loss(self):
        epochs = np.linspace(0, self.nepochs, self.nepochs)
        plt.plot(epochs, self.loss)
        plt.xlabel("Epochs")
        plt.ylabel("Loss")
        plt.show()
